File: Sources/App/Classes/Views/Channel View/TVCLogViewInternalWK2.m
/* *********************************************************************
 *                  _____         _               _
 *                 |_   _|____  _| |_ _   _  __ _| |
 *                   | |/ _ \ \/ / __| | | |/ _` | |
 *                   | |  __/>  <| |_| |_| | (_| | |
 *                   |_|\___/_/\_\\__|\__,_|\__,_|_|
 *
 * Copyright (c) 2010 - 2020 Codeux Software, LLC & respective contributors.
 *       Please see Acknowledgements.pdf for additional information.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *  * Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *  * Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *  * Neither the name of Textual, "Codeux Software, LLC", nor the
 *    names of its contributors may be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *********************************************************************** */

#import "WKWebViewPrivate.h"

#import "IRCChannel.h"
#import "TPCPreferencesLocal.h"
#import "TVCLogController.h"
#import "TVCLogPolicyPrivate.h"
#import "TVCLogScriptEventSinkPrivate.h"
#import "TVCLogViewInternalWK2.h"
#import "TVCLogViewPrivate.h"

NS_ASSUME_NONNULL_BEGIN

@interface TVCLogViewInternalWK2 ()
@property(nonatomic, assign) BOOL t_observingLoadingProperty;
@end

@implementation TVCLogViewInternalWK2

static WKUserContentController *_sharedUserContentController = nil;
static WKWebViewConfiguration *_sharedWebViewConfiguration = nil;
static TVCLogPolicy *_sharedWebPolicy = nil;
static TVCLogScriptEventSink *_sharedWebViewScriptSink = nil;

#pragma mark -
#pragma mark Factory

+ (void)_t_initialize {
  static dispatch_once_t onceToken;

  dispatch_once(&onceToken, ^{
    _sharedWebViewConfiguration = [WKWebViewConfiguration new];

    /* WebKit private API, declared in WKWebViewPrivate.h.
     The style is loaded from a file URL and references other files
     (scripts, images, the theme's own resources) by file URL. WebKit
     treats every file URL as its own origin, so without these two
     flags those references are blocked. There is no public API for
     this; the only alternative is a custom URL scheme handler. The
     remaining private call is -_webView:contextMenu:forElement: in
     the delegate section of this file. */
    _sharedWebViewConfiguration._allowUniversalAccessFromFileURLs = YES;

    _sharedWebViewConfiguration.preferences._allowFileAccessFromFileURLs = YES;

    _sharedWebViewScriptSink =
        [[TVCLogScriptEventSink alloc] initWithWebView:nil];

    _sharedUserContentController = [WKUserContentController new];

    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"appearance"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"channelIsActive"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"channelMemberCount"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"channelName"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"channelNameDoubleClicked"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"displayContextMenu"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"copySelectionWhenPermitted"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"encryptionAuthenticateUser"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"inlineMediaEnabledForView"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"loadInlineMedia"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"localUserHostmask"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"localUserNickname"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"logToConsole"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"networkName"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"nicknameColorStyleHash"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"nicknameDoubleClicked"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"notifyLinesAddedToView"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"notifyLinesRemovedFromView"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"notifyJumpToLineCallback"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"printDebugInformation"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"printDebugInformationToConsole"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"renderMessagesBefore"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"renderMessagesAfter"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"renderMessagesInRange"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"renderMessageWithSiblings"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"renderTemplate"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"retrievePreferencesWithMethodName"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"sendPluginPayload"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"serverAddress"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"serverChannelCount"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"serverIsConnected"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"setChannelName"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"setNickname"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"setLineContext"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"setSelection"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"setURLAddress"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"sidebarInversionIsEnabled"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"styleSettingsRetrieveValue"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"styleSettingsSetValue"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"topicBarDoubleClicked"];
    [_sharedUserContentController
        addScriptMessageHandler:(id)_sharedWebViewScriptSink
                           name:@"finishedLayingOutView"];

    _sharedWebViewConfiguration.userContentController =
        _sharedUserContentController;

    _sharedWebPolicy = [TVCLogPolicy new];
  });
}

- (instancetype)initWithHostView:(TVCLogView *)hostView {
  [self.class _t_initialize];

  if ((self = [self initWithFrame:NSZeroRect
                    configuration:_sharedWebViewConfiguration])) {
    [self constructWebViewWithHostView:hostView];

    return self;
  }

  return nil;
}

- (void)constructWebViewWithHostView:(TVCLogView *)hostView {
  NSParameterAssert(hostView != nil);

  self.t_parentView = hostView;

  self.allowsBackForwardNavigationGestures = NO;
  self.allowsMagnification = NO;

  self.translatesAutoresizingMaskIntoConstraints = NO;

  self.allowsLinkPreview = [TPCPreferences webKit2PreviewLinks];

  self.underPageBackgroundColor = [NSColor clearColor];

  self.customUserAgent = TVCLogViewCommonUserAgentString;

  /* Makes the view show up in Safari's Develop menu and enables the
   "Inspect Element" context menu item. */
  self.inspectable = YES;

  self.navigationDelegate = (id)self;

  self.UIDelegate = (id)self;
}

- (void)dealloc {
  [self stopObservingLoadingProperty];

  self.navigationDelegate = nil;

  self.UIDelegate = nil;
}

- (TVCLogPolicy *)webViewPolicy {
  return _sharedWebPolicy;
}

#pragma mark -
#pragma mark Load Overrides

- (nullable WKNavigation *)loadFileURL:(NSURL *)URL
               allowingReadAccessToURL:(NSURL *)readAccessURL {
  [self startObservingLoadingProperty];

  return [super loadFileURL:URL allowingReadAccessToURL:readAccessURL];
}

- (nullable WKNavigation *)loadHTMLString:(NSString *)string
                                  baseURL:(nullable NSURL *)baseURL {
  [self startObservingLoadingProperty];

  return [super loadHTMLString:string baseURL:baseURL];
}

- (void)stopLoading {
  [self stopObservingLoadingProperty];

  [super stopLoading];
}

#pragma mark -
#pragma mark View Events

- (void)keyDown:(NSEvent *)e {
  if ([self.t_parentView keyDown:e inView:self]) {
    return;
  }

  [super keyDown:e];
}

- (BOOL)performDragOperation:(id<NSDraggingInfo>)sender {
  return [self.t_parentView performDragOperation:sender];
}

#pragma mark -
#pragma mark Utilities

+ (void)emptyCaches {
  WKWebsiteDataStore *wk2WebsiteDataStore =
      _sharedWebViewConfiguration.websiteDataStore;

  if (wk2WebsiteDataStore == nil) {
    return;
  }

  NSSet *itemsToRemove = [NSSet setWithArray:@[
    WKWebsiteDataTypeDiskCache, WKWebsiteDataTypeMemoryCache
  ]];

  [wk2WebsiteDataStore removeDataOfTypes:itemsToRemove
                           modifiedSince:[NSDate distantPast]
                       completionHandler:^{
                         LogToConsoleDebug("WebKit2 cache cleared");
                       }];
}

- (void)findString:(NSString *)searchString movingForward:(BOOL)movingForward {
  NSParameterAssert(searchString != nil);

  WKFindConfiguration *configuration = [WKFindConfiguration new];

  configuration.backwards = (movingForward == NO);
  configuration.caseSensitive = NO;
  configuration.wraps = YES;

  [self findString:searchString
      withConfiguration:configuration
      completionHandler:^(WKFindResult *result) {
        if (result.matchFound == NO) {
          NSBeep();
        }
      }];
}

- (void)startObservingLoadingProperty {
  if (self.t_observingLoadingProperty) {
    return;
  }

  [self addObserver:self
         forKeyPath:@"loading"
            options:NSKeyValueObservingOptionNew
            context:NULL];

  self.t_observingLoadingProperty = YES;
}

- (void)stopObservingLoadingProperty {
  if (self.t_observingLoadingProperty == NO) {
    return;
  }

  [self removeObserver:self forKeyPath:@"loading"];

  self.t_observingLoadingProperty = NO;
}

- (void)maybeInformDelegateWebViewFinishedLoading {
  if (self.t_viewIsLoading == NO && self.t_viewIsNavigating == NO) {
    [self stopObservingLoadingProperty];

    [self.t_parentView performSelectorInCommonModes:@selector
                       (informDelegateWebViewFinishedLoading)
                                         withObject:nil
                                         afterDelay:1.2];
  }
}

- (void)webViewClosedUnexpectedly {
  [self.t_parentView informDelegateWebViewClosedUnexpectedly];
}

#pragma mark -
#pragma mark View Configuration

- (BOOL)maintainsInactiveSelection {
  return YES;
}

#pragma mark -
#pragma mark JavaScript

- (void)logEvaluateJavaScriptError:(NSError *)error {
  NSParameterAssert(error != nil);

  NSNumber *lineNumber = error.userInfo[@"WKJavaScriptExceptionLineNumber"];
  NSString *errorMessage = error.userInfo[@"WKJavaScriptExceptionMessage"];
  NSURL *sourceURL = error.userInfo[@"WKJavaScriptExceptionSourceURL"];

  NSString *channelName =
      self.t_parentView.viewController.associatedChannel.name;

  if (channelName == nil) {
    channelName = @"Server Console";
  }

  if (lineNumber == nil || errorMessage == nil || sourceURL == nil) {
    LogToConsoleError("JavaScript Error in %{private}@: %{public}@",
                      channelName, error.localizedDescription);

    return;
  }

  LogToConsoleError("A JavaScript error occurred in %{public}@ on line "
                    "%{public}ld of %{public}@: %{public}@",
                    channelName, lineNumber.unsignedIntegerValue,
                    sourceURL.standardizedTildePath, errorMessage);
}

- (void)_t_evaluateJavaScript:(NSString *)code
            completionHandler:
                (void (^_Nullable)(id _Nullable))completionHandler {
  NSParameterAssert(code != nil);

  [self evaluateJavaScript:code
         completionHandler:^(id result, NSError *error) {
           if (error) {
             [self logEvaluateJavaScriptError:error];
           }

           if (result) {
             /* WKWebView reports JavaScript "undefined" as a nil result and
   "null" as NSNull; the legacy WebUndefined never appears. */
             if ([result isKindOfClass:[NSNull class]]) {
               if (completionHandler) {
                 completionHandler(nil);
               }

               return;
             }
           }

           if (completionHandler) {
             completionHandler(result);
           }
         }];
}

#pragma mark -
#pragma mark Web View Delegate

- (void)webViewWebContentProcessDidTerminate:(WKWebView *)webView {
  NSParameterAssert(webView == self);

  /* The content process is gone (crash, memory pressure, jetsam).
   The view is blank until something is loaded into it again, so
   hand off to the controller which reloads the document and replays
   the history. */
  LogToConsoleError(
      "WebView [%{public}@] content process terminated. Reloading view.",
      self.description);

  self.t_viewIsLoading = NO;
  self.t_viewIsNavigating = NO;

  [self stopObservingLoadingProperty];

  [self webViewClosedUnexpectedly];
}

- (void)observeValueForKeyPath:(nullable NSString *)keyPath
                      ofObject:(nullable id)object
                        change:(nullable NSDictionary<NSString *, id> *)change
                       context:(nullable void *)context {
  NSParameterAssert(object == self);

  if ([keyPath isEqualToString:@"loading"]) {
    self.t_viewIsLoading = self.loading;

    [self maybeInformDelegateWebViewFinishedLoading];
  }
}

- (void)webView:(WKWebView *)webView
    decidePolicyForNavigationAction:(WKNavigationAction *)navigationAction
                    decisionHandler:
                        (void (^)(WKNavigationActionPolicy))decisionHandler {
  NSParameterAssert(webView == self);

  [_sharedWebPolicy webView2:webView
                              logView:self.t_parentView
      decidePolicyForNavigationAction:navigationAction
                      decisionHandler:decisionHandler];
}

- (void)webView:(WKWebView *)webView
    didReceiveAuthenticationChallenge:(NSURLAuthenticationChallenge *)challenge
                    completionHandler:
                        (void (^)(NSURLSessionAuthChallengeDisposition,
                                  NSURLCredential *))completionHandler {
  NSParameterAssert(webView == self);

  [_sharedWebPolicy webView2:webView
                                logView:self.t_parentView
      didReceiveAuthenticationChallenge:challenge
                      completionHandler:completionHandler];
}

- (void)webView:(WKWebView *)webView
    didStartProvisionalNavigation:(WKNavigation *)navigation {
  NSParameterAssert(webView == self);

  self.t_viewIsNavigating = YES;
}

- (void)webView:(WKWebView *)webView
    didFinishNavigation:(WKNavigation *)navigation {
  NSParameterAssert(webView == self);

  self.t_viewIsNavigating = NO;

  [self maybeInformDelegateWebViewFinishedLoading];
}

/* WebKit private API (WKUIDelegatePrivate). WKWebView offers no public
 hook for replacing the context menu while keeping WebKit's own items
 (Inspect Element, Look Up, Search With Google), so this is the one
 remaining private delegate call. Should WebKit stop calling it, the
 view falls back to WebKit's default menu. */
- (NSMenu *)_webView:(WKWebView *)webView
         contextMenu:(NSMenu *)menu
          forElement:(id)element {
  NSParameterAssert(webView == self);

  return [_sharedWebPolicy webView2:webView
                            logView:self.t_parentView
         contextMenuWithDefaultMenu:menu];
}

@end

NS_ASSUME_NONNULL_END

File: Sources/App/Classes/Views/Channel View/TVCLogViewInternalWK2.swift
/* *********************************************************************
 *                  _____         _               _
 *                 |_   _|____  _| |_ _   _  __ _| |
 *                   | |/ _ \ \/ / __| | | |/ _` | |
 *                   | |  __/>  <| |_| |_| | (_| | |
 *                   |_|\___/_/\_\\__|\__,_|\__,_|_|
 *
 * Copyright (c) 2010 - 2026 Codeux Software, LLC & respective contributors.
 *       Please see Acknowledgements.pdf for additional information.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *  * Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *  * Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *  * Neither the name of Textual, "Codeux Software, LLC", nor the
 *    names of its contributors may be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *********************************************************************** */

import AppKit
import os
import WebKit

private let logViewWebKitLogger = Logger(
	subsystem: Bundle.main.bundleIdentifier ?? "Glasstual",
	category: "LogViewWebKit"
)

@objc(TVCLogViewInternalWK2)
@MainActor
final class LogViewWebView: WKWebView, WKNavigationDelegate, WKUIDelegate {
	/** One configuration, one user content controller, one script sink and one
	 policy for every channel, on purpose. The sink resolves the sender from
	 `message.webView`, so it needs no per-view registration, and because no
	 view owns a handler the usual advice to remove handlers in `deinit` does
	 not apply here. The cost is one process pool and one website data store
	 shared by every log view. */
	@MainActor
	private enum SharedResources {
		/** Owned by the sink so a registered name without a handler cannot
		 exist. `encryptionAuthenticateUser` was registered here and
		 implemented nowhere; it is gone along with the rest of OTR. */
		static let messageHandlerNames = TVCLogScriptEventSink.registeredMessageNames

		static let scriptSink = TVCLogScriptEventSink(webView: nil)
		static let policy = LogPolicy()
		static let configuration: WKWebViewConfiguration = {
			let configuration = WKWebViewConfiguration()

			/* The document and every theme resource it references are served
			 by one scheme handler, so the page has a single origin and no
			 file-system read access. */
			configuration.setURLSchemeHandler(
				LogViewThemeSchemeHandler.shared,
				forURLScheme: LogViewContentPolicy.themeScheme
			)

			let contentController = WKUserContentController()
			for name in messageHandlerNames {
				contentController.add(scriptSink, name: name)
			}
			configuration.userContentController = contentController
			return configuration
		}()
	}

	@objc(t_parentView) weak var parentView: LogView?
	@objc(t_viewIsLoading) var viewIsLoading = false
	@objc(t_viewIsNavigating) var viewIsNavigating = false

	private var loadingObservation: NSKeyValueObservation?

	init() {
		super.init(frame: .zero, configuration: SharedResources.configuration)
	}

	@available(*, unavailable)
	required init?(coder _: NSCoder) {
		fatalError("init(coder:) has not been implemented")
	}

	@objc(initWithHostView:)
	convenience init(hostView: LogView) {
		self.init()
		attach(to: hostView)
	}

	func attach(to hostView: LogView) {
		parentView = hostView
		allowsBackForwardNavigationGestures = false
		allowsMagnification = false
		translatesAutoresizingMaskIntoConstraints = false
		allowsLinkPreview = TextualPreferences.webKit2PreviewLinks()
		underPageBackgroundColor = .clear
		customUserAgent = LogView.commonUserAgent
		/* The log view holds a native bridge and runs third-party theme
		 JavaScript; the inspector is a development tool, not a shipped one. */
		#if DEBUG
			isInspectable = true
		#endif
		navigationDelegate = self
		uiDelegate = self
	}

	deinit {
		loadingObservation?.invalidate()
	}

	@objc var webViewPolicy: LogPolicy {
		SharedResources.policy
	}

	@discardableResult
	override func load(_ request: URLRequest) -> WKNavigation? {
		startObservingLoading()
		return super.load(request)
	}

	override func stopLoading() {
		stopObservingLoading()
		super.stopLoading()
	}

	override func keyDown(with event: NSEvent) {
		if parentView?.keyDown(event, in: self) == true {
			return
		}
		super.keyDown(with: event)
	}

	override func performDragOperation(_ sender: NSDraggingInfo) -> Bool {
		parentView?.performDragOperation(sender) ?? false
	}

	@objc static func emptyCaches() {
		let types: Set<String> = [WKWebsiteDataTypeDiskCache, WKWebsiteDataTypeMemoryCache]
		SharedResources.configuration.websiteDataStore.removeData(
			ofTypes: types,
			modifiedSince: .distantPast
		) {
			logViewWebKitLogger.debug("WebKit cache cleared")
		}
	}

	@objc(findString:movingForward:)
	func find(_ searchString: String, movingForward: Bool) {
		let configuration = WKFindConfiguration()
		configuration.backwards = movingForward == false
		configuration.caseSensitive = false
		configuration.wraps = true
		find(searchString, configuration: configuration) { result in
			if result.matchFound == false {
				NSSound.beep()
			}
		}
	}

	private func startObservingLoading() {
		guard loadingObservation == nil else {
			return
		}

		/* No .initial: the observation is installed before the load starts, so
		 the initial callback reports isLoading == false and would declare the
		 view finished against a blank page. */
		loadingObservation = observe(\.isLoading, options: [.new]) { [weak self] _, change in
			MainActor.assumeIsolated {
				guard let self else {
					return
				}
				self.viewIsLoading = change.newValue ?? self.isLoading
				self.maybeInformDelegateWebViewFinishedLoading()
			}
		}
	}

	private func stopObservingLoading() {
		loadingObservation?.invalidate()
		loadingObservation = nil
	}

	private func maybeInformDelegateWebViewFinishedLoading() {
		guard viewIsLoading == false, viewIsNavigating == false else {
			return
		}
		stopObservingLoading()
		parentView?.perform(
			#selector(LogView.informDelegateWebViewFinishedLoading),
			with: nil,
			afterDelay: 1.2,
			inModes: [.common]
		)
	}

	@objc var maintainsInactiveSelection: Bool {
		true
	}

	@objc(_t_evaluateJavaScript:completionHandler:)
	func evaluate(_ code: String, completionHandler: ((Any?) -> Void)?) {
		evaluateJavaScript(code) { [weak self] result, error in
			if let error {
				self?.logJavaScriptError(error)
			}
			completionHandler?(result is NSNull ? nil : result)
		}
	}

	/** Runs `body` with `arguments` bound by name. WebKit converts the values,
	 which removes every hand-written escape from the bridge. */
	func call(_ body: String, arguments: [String: Any]) async throws -> Any? {
		do {
			let result = try await callAsyncJavaScript(
				body,
				arguments: arguments,
				in: nil,
				contentWorld: .page
			)
			return result is NSNull ? nil : result
		} catch {
			logJavaScriptError(error)
			throw error
		}
	}

	private func logJavaScriptError(_ error: Error) {
		let nsError = error as NSError
		let lineNumber = nsError.userInfo["WKJavaScriptExceptionLineNumber"] as? NSNumber
		let errorMessage = nsError.userInfo["WKJavaScriptExceptionMessage"] as? String
		let sourceURL = nsError.userInfo["WKJavaScriptExceptionSourceURL"] as? URL
		let channelName = parentView?.viewController?.associatedChannel?.name ?? "Server Console"

		guard let lineNumber, let errorMessage, let sourceURL else {
			logViewWebKitLogger.error(
				"JavaScript error in \(channelName, privacy: .private): \(error.localizedDescription, privacy: .public)"
			)
			return
		}

		let sourcePath = sourceURL.standardizedFileURL.path(percentEncoded: false)
		let errorDescription = "JavaScript error in \(channelName) on line \(lineNumber.uintValue) of \(sourcePath): \(errorMessage)"
		logViewWebKitLogger.error("\(errorDescription, privacy: .public)")
	}

	func webViewWebContentProcessDidTerminate(_ webView: WKWebView) {
		precondition(webView === self)
		let viewDescription = description
		logViewWebKitLogger.error("WebView content process terminated. Reloading \(viewDescription, privacy: .public)")
		viewIsLoading = false
		viewIsNavigating = false
		stopObservingLoading()
		parentView?.informDelegateWebViewClosedUnexpectedly()
	}

	func webView(
		_ webView: WKWebView,
		decidePolicyFor navigationAction: WKNavigationAction,
		decisionHandler: @escaping @MainActor (WKNavigationActionPolicy) -> Void
	) {
		precondition(webView === self)
		guard let parentView else {
			decisionHandler(.cancel)
			return
		}

		SharedResources.policy.webView2(
			webView,
			logView: parentView,
			decidePolicyFor: navigationAction,
			decisionHandler: decisionHandler
		)
	}

	func webView(
		_ webView: WKWebView,
		didReceive challenge: URLAuthenticationChallenge,
		completionHandler: @escaping @MainActor (URLSession.AuthChallengeDisposition, URLCredential?) -> Void
	) {
		precondition(webView === self)
		guard let parentView else {
			completionHandler(.performDefaultHandling, nil)
			return
		}

		SharedResources.policy.webView2(
			webView,
			logView: parentView,
			didReceive: challenge,
			completionHandler: completionHandler
		)
	}

	func webView(_ webView: WKWebView, didStartProvisionalNavigation _: WKNavigation!) {
		precondition(webView === self)
		viewIsNavigating = true
	}

	func webView(_ webView: WKWebView, didFinish _: WKNavigation!) {
		precondition(webView === self)
		viewIsNavigating = false
		maybeInformDelegateWebViewFinishedLoading()
	}

	/** WKWebView has no public hook that replaces the context menu while
	 retaining WebKit's Inspect Element, Look Up, and Search items. If
	 WebKit stops invoking this private delegate method, its default menu
	 remains available. */
	@objc(_webView:contextMenu:forElement:)
	func webView(_ webView: WKWebView, contextMenu menu: NSMenu, forElement _: Any) -> NSMenu {
		precondition(webView === self)
		guard let parentView else {
			return menu
		}

		return SharedResources.policy.webView2(
			webView,
			logView: parentView,
			contextMenuWithDefaultMenu: menu
		)
	}
}
